### PR TITLE
Support JupyterLab 2, add jupyterlab-go-to-definition to monorepo

### LINKED ISSUE #186: Migrate to JupyterLab 2.0
We will need to branch off at some time in order to migrate to JupyterLab 2.0 (which will be released soon). It would be good to merge any existing PRs (other than my stale ones) and avoid starting any new substantial work (other than bugfixes), particularly anything touching the JupyterLab internals. Also, the icons system is updated in 2.0 - that's why I did not add a nicer icon so far.

There is a draft migration guide here: https://jupyterlab.readthedocs.io/en/latest/developer/extension_migration.html. If we have time to do this early, we can contribute to making it more complete.

I might have some time this weekend but will rather spend it writing more browser tests so we can have a well-tested transition. @bollwyvl please let me know if you have any other local work related to browser tests that I should wait for.

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- fixes #186

## Code changes

- [x] add jupyterlab-go-to-definition
  - [x] pass lint
  - [x] get compiling
  - [x] get tests running
- [x] bump versions
- [x] get everything compiling
- [x] get lab at least building
- [x] get working on binder
  - [Binder](https://mybinder.org/v2/gh/bollwyvl/jupyterlab-lsp/add-lab2?urlpath=lab%2Ftree%2Fexamples%2FPython.ipynb)
- [x] get acceptance tests working
  - [x] fix selectors
  - [x] turn off build check during acceptance test
- [x] get js unit tests working
- [x] update docs
  - [x] needs node 10, now, apparently
  - [x] do another pass through installation, etc.

Not on this PR:
- add initial integration with _Running_ sidebar #205

## User-facing changes

- will install in JupyterLab 2.0

## Backwards-incompatible changes

- shouldn't be much that will be hard to back/forward port, but this is a pretty hard break

## Chores

- [x] linted
- [x] tested
- [x] documented
- [x] changelog entry

### CODE REVIEW

**Comment by bollwyvl** [comment]
About an hour and a half in, I've presently gotten the main `tsc` errors down to:
```
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/components/completion.ts(13,10): error TS2305: Module '"../../../../../../node_modules/@jupyterlab/apputils/lib"' has no exported member 'IClientSession'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/codemirror/features/diagnostics_listing.tsx(250,5): error TS2554: Expected 1 arguments, but got 0.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/codemirror/features/hover.ts(10,10): error TS2305: Module '"../../../../node_modules/@jupyterlab/coreutils/lib"' has no exported member 'Debouncer'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/file_editor.ts(90,7): error TS2741: Property 'list' is missing in type 'LSPConnector' but required in type 'IDataConnector<IReply, void, IRequest, string>'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(18,10): error TS2305: Module '"../../../../../node_modules/@jupyterlab/apputils/lib"' has no exported member 'IClientSession'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(52,25): error TS2339: Property 'session' does not exist on type 'IContext<INotebookModel>'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(60,21): error TS2694: Namespace '"/home/weg/projects/lsp/jupyterlab-lsp/node_modules/@jupyterlab/services/lib/session/session"' has no exported member 'IKernelChangedArgs'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(67,19): error TS7006: Parameter 'spec' implicitly has an 'any' type.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(76,14): error TS7006: Parameter 'e' implicitly has an 'any' type.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(87,25): error TS2339: Property 'session' does not exist on type 'IContext<INotebookModel>'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(106,27): error TS2339: Property 'session' does not exist on type 'IContext<INotebookModel>'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(117,34): error TS2339: Property 'session' does not exist on type 'IContext<INotebookModel>'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(176,28): error TS2339: Property 'session' does not exist on type 'NotebookPanel'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(190,7): error TS2322: Type 'LSPConnector' is not assignable to type 'IDataConnector<IReply, void, IRequest, string>'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/notebook.ts(204,5): error TS2322: Type 'LSPConnector' is not assignable to type 'IDataConnector<IReply, void, IRequest, string>'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/components/statusbar.tsx(20,10): error TS2305: Module '"../../../../../../node_modules/@jupyterlab/ui-components/lib"' has no exported member 'DefaultIconReact'.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/components/statusbar.tsx(94,5): error TS2554: Expected 1 arguments, but got 0.
@krassowski/jupyterlab-lsp-metapackage: ../jupyterlab-lsp/src/adapters/jupyterlab/components/statusbar.tsx(233,5): error TS2554: Expected 1 arguments, but got 0.
```

Not too bad!

**Comment by bollwyvl** [comment]
File editors seem to mostly work. Notebook still seems broken...

**Comment by bollwyvl** [comment]
Some of the things on notebook (which anecdotally works locally):
- we get the sessionContext _much_ earlier, and so we get the _first_ kernel change (which we're already waiting for) and everything was initializing twice
- the codemirror-based mouse movement api changed (from `true/false` to `1/-1`)
- some null checks weren't firing on `undefined`
- a few other lifecycle things were happening far earlier, such as diagnostics... just dropping all diagnostics on a zero-length document on the floor without complaining... we'd have to add a promise or something to know when it's initialized for the first time
- when closing a document, things get disposed/reset by various parties, before we're notified on the signal... it's hard to dig out we're in 

All that aside, and despite the continuing CI failing (looks like nodejs 8 might cut it anymore), it looks like binder can be ticked off///

**Comment by bollwyvl** [comment]
The jest stuff was blessedly very short to fix, works locally, but would like to see a CI pass before marking it done.

**Comment by bollwyvl** [comment]
I have to put this down for the day, but we have our first green ci run with 27c882b :tada: so i think this is ready to review. that last push adds the build check skip, which _might_ help with windows flake, since the machine won't be working so hard doing node stuff.

i guess we can't land this until whatever happens with `0.8` happens, and it will take some release engineering to get this out, but it would be great to offer this up for folks.

File: .travis.yml [hunk 1]
```diff
@@ -6,17 +6,24 @@ cache:
...
+    - TRAVIS_NODE_VERSION=10
+    - TRAVIS_NODE_VERSION=13
+
 matrix:
   include:
     - python: '3.5'
-    - python: '3.6'
-    - python: '3.7'
-    - python: '3.8'
+    - python: '3.9-dev'
```
Comment (krassowski) [inline .travis.yml]: I would keep 3.8.

Comment (bollwyvl) [reply .travis.yml]: hopefully addressed by bdd488c and 1345269!

File: atest/Keywords.robot [hunk 2]
```diff
@@ -33,6 +30,16 @@ Setup Server and Browser
     Set Global Variable    ${PAGE CONFIG}    ${config}
     Set Global Variable    ${LAB VERSION}    ${config["appVersion"]}
 
+Create Lab Launch Command
+    [Arguments]    ${root}
+    ${WORKSPACES DIR} =    Set Variable    ${OUTPUT DIR}${/}workspaces
+    ${app args} =    Set Variable    --no-browser --debug --NotebookApp.base_url\='${BASE}' --port\=${PORT} --NotebookApp.token\='${TOKEN}'
+    ${path args} =    Set Variable    --LabApp.user_settings_dir='${SETTINGS DIR.replace('\\', '\\\\')}' --LabApp.workspaces_dir\='${WORKSPACES DIR.replace('\\', '\\\\')}'
+    ${ext args} =    Set Variable    --LanguageServerManager.extra_node_roots\="['${root.replace('\\', '\\\\')}']"
+    ${build args} =    Set Variable    --LabApp.tornado_settings\='{"page_config_data": {"buildCheck": 0, "buildAvailable": 0}}'
```
Comment (bollwyvl) [inline atest/Keywords.robot]: I don't know what i wasn't escaping properly, but i think the fixture-based approach will be fine. Future options would be doing a python-based keyword, like `ports.py` or perhaps a templated fixture: shouldn't block this PR, I'd think.

**Comment by bollwyvl** [comment]
The general rationale on nodejs support change: The current setup gives us some coverage across versions. 8 is out of its [support window](https://nodejs.org/en/about/releases/). I would also be tempted to add a JupyterLab build at the end of each of them, but that would slow down our "fast" tests a great deal. We could also template out across the node versions on azure, which may be worth doing, but maybe not on this PR.

**Comment by bollwyvl** [comment]
Thanks for the fix on that template typo: probably need to hoist that matrix at some point so it's not repeated everywhere, but hooray green!

File: atest/Keywords.robot (hunk 2 above)
Comment (bollwyvl) [reply atest/Keywords.robot]: regarding azure node matrixing: after a rough start on f033e5f, finished on 39acc4c ... didn't uncover any issues (even on windows, which was notoriously bad in the past regarding long paths) but good to know anyway.

**Comment by bollwyvl** [comment]
I don't anticipate doing any other changes here... though i might start the follow-on tinkering for #205, potentially just adding/testing the backend routes first...

**Comment by krassowski** [comment]
Sorry for the delay here. I will merge #206 within days, release 0.8 and move onto 2.0!

**Comment by bollwyvl** [comment]
No worries! Good luck on the extractors!

**Comment by bollwyvl** [comment]
Hooray, back to green!

**Comment by krassowski** [comment]
I branched out [jupyterlab-lsp/tree/0.8.x](https://github.com/krassowski/jupyterlab-lsp/tree/0.8.x) (so we can continue supporting JL 1.x there) and will merge this PR into master now.

**Pull request merged**